### DISCUSSION TITLE
Prisma, Next.js, Heroku PostgreSQL database running into connection limits in production

### DISCUSSION THREAD

**Discussion by 0x6A6471** [prisma/prisma#6913] [Q&A] [answered]
It seems a lot of people have different issues surrounding this topic.  Either they are unanswered or proposed solutions don't seem to work, so I figured I'd ask for help here. 

In development, when monitoring my Heroku connections I have a single connection (which is expected), but in Vercel’s preview environment (production) my connections skyrocket when using Prisma. I’m currently only instantiating a single PrismaClient like the docs say, and implementing a `connection_limit` doesn't do anything. 

My Prisma instantiation is based off of this article: (https://www.prisma.io/docs/support/help-articles/nextjs-prisma-client-dev-practices)

**Comment by janpio** [comment, marked as answer]
Vercel deploys as serverless functions, which means each environment and deployment will use its own connection. Then run 2 requests in parallel, 2 connections are in use. Each new deployments get a completely new runtimes - and all the open connections from before will run until AWS shuts down these running functions (which takes 10 to 120 minutes usually)

Real workarounds for this behavior does not exist - and it can not be influenced by Prisma in any way as the old environments are technically frozen - no code is running that could e.g. close your connection after 2 minutes. 

So what you can do:
- Put a connection pooler in front of your database so it matters later or not at all.
- You can manually kill all connections from time to time to close the connections to the old function runtimes.
- You could also write some code that does this automatically every x minutes or on demand.
- Or you can configure your database to automatically close connections after x minutes of inactivity.

**Reply by 0x6A6471** [reply]
Okay, thank you. I will look into connection pooling with PgBouncer.

**Reply by 0x6A6471** [reply]
Just an update for anyone with similar issues, setting up a connection pool and using PgBouncer works as wanted.

**Reply by lovegrovegeorge** [reply]
@jakequinter it looks like I may need to go down this route as well sir! Did you find any useful resources on setting that all up that you could share? 🙏

**Reply by 0x6A6471** [reply]
Hi @georgelovegrove, I've been meaning to write up a blog post about it, but haven't gotten that far yet.  In the meantime, this [YouTube video](https://www.youtube.com/watch?v=tsOHw2T1HvI&ab_channel=Prisma) seemed to help from Prisma.

**Reply by lovegrovegeorge** [reply]
Ah fantastic I'll take a look, if you do make any post share on here and will eagerly review. I'll also share anything I find to help others who will need to do this!

**Reply by lovegrovegeorge** [reply]
Hey @jakequinter I took a look at the video and some resources. Did you go down the digital ocean route? That setup is nice and easy for the connection pooling. I'm trying to use Vercel and their serverless functions so want to host DB with AWS in the same AWS region as Vercel uses. Not seeing any great resource on setting up pgbouncer in AWS in EC2 instances, worth me checking what you went with!

**Reply by 0x6A6471** [reply]
Yeah, I ended up migrating to a Digital Ocean PostgreSQL instance for the time being. Mainly for testing purposes, but it was painless (for the most part) to get it up and working quickly.  I may look into alternatives for production use cases when the time comes, but Digital Ocean seems to be working wonderfully right now.

**Reply by lovegrovegeorge** [reply]
Yeh looked like a fairly easy setup with Digital Ocean. I'm going to kick the can down the road and not use a connection pooler for the minute due to having low scale requirements for now. The pgbouncer setup adds a learning and maintenance overhead I'd want to be certain is worth the time investment, can decide in a few months. From what I can see Prisma team may revisit it soon (https://github.com/prisma/prisma/issues/5866). Hopefully others highlight the issue for themselves and the priority increases over time for good RDS proxy integration. That or I may get a realtime subscription requirement and just move from serverless to a full server, time will tell 🤔

**Reply by janpio** [reply]
(Heads up: Do not get your hopes up to much about AWS RDS Proxy soon, Prisma is really incompatible with it in a bad fundamental way that we will have to invest a _lot_ of effort to work around.)

**Reply by lovegrovegeorge** [reply]
@janpio cuts me deep haha appreciate the intel though as always!
